### DISCUSSION TITLE
Filter a language typed string & use it without language tag

### DISCUSSION THREAD

**Discussion by ktk** [eyereasoner/eye#68] [Q&A] [answered]
I have a predicate with literals in 4 languages. I would like to select just one language and attach it to a `rdfs:label` property without any language tag. Kind of the "default" label to use.

By going through the built-ins I guess this might be `pred:matches-language-range` but I would definitely need an example for that.

Anyone can enlighten me?

**Comment by josd** [comment]
RIF examples and test cases are at https://github.com/eyereasoner/eye/tree/master/reasoning/rif
Your specific example is at https://github.com/eyereasoner/eye/blob/master/reasoning/rif/rifP.n3#L382

For other built-ins see also https://github.com/eyereasoner/eye/tree/master/reasoning/bi and https://github.com/eyereasoner/Notation3-By-Example

The N3 Comminity Group is currently working on extensive N3 built-ins documentation.

**Reply by ktk** [reply]
Yes I saw it mentioned in https://github.com/eyereasoner/eye/discussions/67, looking forward!

**Comment by phochste** [comment, marked as answer]
The built-in `log:langlit` could also be used.

```
@prefix : <urn:example.org:> .
@prefix log: <http://www.w3.org/2000/10/swap/log#> .

:Cat a :Animal ;
    :label "cat"@en , "chat"@fr , "kat"@nl, "kucing"@id .

{
    ?Subject :label ?Label .
    ( ?LabelText "id" ) log:langlit ?Label .
}
=>
{
    ?Subject :lang ?LabelText .
} .
```

which gives  after running `eye --pass --nope --quiet test.n3` :

```
@prefix : <urn:example.org:>.
@prefix log: <http://www.w3.org/2000/10/swap/log#>.

<urn:example.org:Cat> a <urn:example.org:Animal>.
<urn:example.org:Cat> <urn:example.org:label> "cat"@en.
<urn:example.org:Cat> <urn:example.org:label> "chat"@fr.
<urn:example.org:Cat> <urn:example.org:label> "kat"@nl.
<urn:example.org:Cat> <urn:example.org:label> "kucing"@id.
<urn:example.org:Cat> <urn:example.org:lang> "kucing".
```

**Reply by ktk** [reply]
Thanks, this works! But I adjusted it wrong first, my (naive?) interpretation of Notation3 rules was that the left part is kind of the argument and the right part is the return value of the function?

But in this example it binds it to `?LabelText` instead of `?Label` like I expected. What are the general rules here, is that explained somewhere?

**Reply by phochste** [reply]
This rules will be explained by the W3C N3 group that is working on rewriting the documentation about it (I'm part of it too).

You need to think of built-ins as a sort of prolog on the web. Many can work in two directions. Eg. `list:append` ..appends things to a list

```
   ( (1) (2 3 4) ) list:append (1 2 3 4 ).    # a true statement

   # What need to be append to (1) to make (1 2 3 4)
   ( (1) ?What ) list:append (1 2 3 4).   # ?What binds to (2 3 4)

   # What needs to be prepend to (2 3 4) to make (1 2 3 4)
   (  ?What (2 3 4) ) list:append (1 2 3 4).   # ?What binds to (1)
   
   # What is (1) with appending (2 3 4)
  (  (1) (2 3 4) ) list:append ?What.   # ?What binds to (1 2 3 4)
```

The new documentation should state in which directions the built-in works.

Some built-ins can also give multiple answers:

```
  ( 1 2 3 4) list:member ?What # Produces 4 different bindings of What : ?What = 1 , ... ?What = 4
```

Some built-ins don't produce any binding when they are unbounded variables (they would provide infinite answers otherwise)

```
    42 math:greaterThan ?What  # If What is not bounded , then this is false ..there would be infinite many numbers bigger than 42.
```

The new documentation should state how many answers a built-in can produce. I hope [Notation3-By-Example](https://github.com/eyereasoner/Notation3-By-Example) can already provide some hints.

**Comment by josd** [comment]
Yes @phochste, this is much better and I would agree to drop the 120 RIF built-ins and only use the 7 core built-in groups

1. `crypto:`
2. `graph:`
3. `list:`
4. `log:`
5. `math:`
6. `string:`
7. `time:`
